### DISCUSSION TITLE
What are the benefits of using Prisma over Mongoose? How do you decide which ORM to use in your projects?

### DISCUSSION THREAD

**Discussion by lumenwrites** [prisma/prisma#9076] [Q&A] [answered]
Hi! I'm relatively new at this, sorry if this is a simple question, but I'm hoping you guys can help me to clear this up.

Up until recently I've only been building apps with Mongoose, but I seem to notice from blogs/tutorials/examples that many people are using Prisma nowadays. I have built a simple project using Prisma to try to experiment with it, and it seemed to work fine, but I still don't have a deep understanding of its benefits, or why should it be preferred over Mongoose.

To me it felt like Prisma was a bit more difficult to use and wrap my head around, a bit slower to develop with, and I ran into some issues that were hard to solve. But I'm novice at this, I'm sure there are reasons why people tend to prefer Prisma, and I'd like to understand them.

I'm starting a pretty big project, and I'm trying to decide between Prisma and Mongoose. I don't know if I have a good reason to prefer one or the other, it seems like either one would work just fine. But Mongoose seems simpler to use and faster to build stuff with, and Prisma seems to be the thing people are switching to (based on what I see on forums).

Could you share your opinion? How do you make decisions when choosing between ORMs? What kind of factors should I consider, what are the pros and cons of Prisma and Mongoose?

I'd like to be able to make this decision intelligently, not based on what's easy and familiar vs what's cool.

(I know that there are other ORMs aside from these two, but these two seem to be more popular from what I see. I figured it's not worth confusing myself even more with more alternatives, since I don't even have a good way to decide between these two.)

Basically, I'm just hoping for some advice from a more experienced developer - how should I think about these things, what criteria do people use to make these choices?

**Comment by matthewmueller** [comment, marked as answer]
Hey Mike, thanks for the great question! Speaking as someone who's working on MongoDB for Prisma, I'll try to leave my bias at the door.

My personal suggestion would be to use what you know. If you're comfortable working with Mongoose, then go with that. It was the first major ORM to be written in Node, so it's had many years to add useful features and squash bugs.

What you may find down the road with Mongoose is that it's easy to lose track of the structure of your database. As you add features and fix bugs, the structure of your data will change. This is best illustrated with an example: 

Maybe your application has zip codes and for the longest time it was just a number, but now you realize that it should actually be a string. If you simply adjust the type from a number to a string in Mongoose and start writing strings to zipCode, Mongoose will merrily do that for you. Without consciously migrating your existing data, your database will have zip codes with both numbers and strings. This may introduce bugs in your application because Mongoose will lead you to believe that zip codes are all strings. This may seem like a simple fix, but these changes compound over the lifetime of an application.

Prisma enforces a schema on top of MongoDB. You can find an example schema in [our documentation](https://www.prisma.io/docs/concepts/components/prisma-schema#example). With a schema, you always know the structure of your database.

We enforce a schema because we believe that:
- Understanding the structure of your data is essential to maintaining a project over a long period of time. This is especially important if you plan to onboard new developers.
- Changing the structure of your data (e.g. number to string) is always an important event. It shouldn't appear to work when it doesn't. We can provide tools like [Prisma Migrate](https://www.prisma.io/migrate) to streamline such events.
- By knowing the schema, you can build sophisticated tools to provide a better developer experience that are tailored to your application. Tools like our type-safe [Prisma Client](https://www.prisma.io/client) and [Prisma Studio](https://www.prisma.io/studio).

MongoDB support in Prisma is still in it's early days, we don't yet have a Migration tool or Studio support. What we do have is a type-safe, auto-generated database client called the Prisma Client. You can get a better sense of what the Prisma Client supports in [our documentation](https://www.prisma.io/docs/concepts/components/prisma-client). It's not a 1:1 mapping with Mongoose, but we support most of the things you need to do in Mongoose in a more type-safe way.

We'll also make it easy to switch from Mongoose to Prisma once we introduce [Introspection Support for MongoDB](https://github.com/prisma/prisma/issues/6787). So you have time to make your decision.

Hope this helps!

**Reply by lumenwrites** [reply]
Thank you so much for such an amazing reply! Very helpful!

**Reply by 2color** [reply]
Hey Mike 👋,
In addition to what Matt wrote, we also have an [open feedback issue](https://github.com/prisma/prisma/issues/8241) where you can share your experience and any stumbling blocks when using Prisma with MongoDB. 

I'm looking forward to hearing what you choose and how it goes.

**Reply by lumenwrites** [reply]
Thanks!

**Reply by raviroshanmehta** [reply]
From last 5 years i was using Mongoose and decided to use Prisma for next big project. 

Initially it was very hard to adopt Prisma but later of i realized it's  type safe property is very useful.  

However I am missing the multiple models creation in Prisma and using below hack.

In prisma folder of project i have created a new folder models to define multiple models prisma files and added below script to sync all definitions. 
```
"scripts": {
    "db": "cat prisma/models/*.prisma > prisma/schema.prisma && npx prisma format && npx prisma db push",
    "seed":"npx prisma db seed"
 },
```

This hack works fine for linux based systems for windows devs in my team can't run this command. 

Looking  forwad to get support for the same from officials prisma to get rid of this hack.

**Comment by ronnostic** [comment]
I stumbled upon this page while comparing the two as well, and after further research, decided to stick with Mongoose, specifically `mongoose-ts` for the following reasons:

* Prisma still has limited support for MongoDB see [this page](https://www.prisma.io/docs/concepts/database-connectors/mongodb) and the tool in general still  appears primarily concerned with relational databases
* Prisma introduces a learning curve, especially for developers who are new to the tool. The Prisma client API and PQL might require time to grasp fully.
* Prisma appears to generate database schemas automatically based on your Prisma schema. While this can be convenient, it may lead to limited control over the database schema, which can be a concern for complex projects with specific schema requirements.
* Mongoose has been around since 2010 

So while it's true that Prisma offers schema enforcement as one of its advantages, a Mongoose, TypeScript, and GraphQL stack on top of MongoDB can mitigate the risk described above in several ways: 
* Using TypeScript allows developers to define strongly typed schemas, resolvers, and GraphQL types. 
* Using Mongoose-TS (or Typegoose) for Schema Definitions allows you to define schemas using TypeScript interfaces or classes with decorators, providing type safety and schema validation.  This provides strong type annotations for schema definitions - you also define the expected types for each field - which TypeScript then enforces - preventing runtime type errors and providing autocompletion and type checking during development as well as the ability to infer types.
* Using GraphQL enforces a clear and structured API with defined types and queries, offering schema-like enforcement at the API level.
* Additionally you could introduce validation libraries like Joi or Yup to add an extra layer of validation to your data at the application level. 
* If you need migration tooling, you have flexible choices here as well like Mongoose-Migrate or Mongeez. 

In addition to all of this, this you have good 'ol version control and documentation. :)  It's important to keep in mind that MongoDB, being a NoSQL database, it is designed for flexibility regarding schemas.  Therefore, not enforcing a strict schema is a feature that makes managing schema changes more flexible compared to relational databases.  I feel that combined with good development practices and the above tooling you can still ensure data consistency and structure while enjoying the flexibility and power of NoSQL

Just my $.02

**Reply by RaschidJFR** [reply]
Was `mongoose-ts` a package? I can't find anything related to it.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
